### PR TITLE
Add selection vectors

### PR DESCRIPTION
**Pull request by Salkin975**
adds base methods for re integration of filter methods

- [x] reach test coverage of over 90% on added files

### CODE REVIEW

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
+import {beforeEach, describe, it, expect} from "vitest";
+import {SequenceSelectionVector} from "./sequenceSelectionVector";
+
+let seqVector: SequenceSelectionVector;
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: It is best to move this and the `beforeEach` into the first (main) `describe` method. This way when reading this it easy to understand that they are "linked".

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+beforeEach(() => {
+    seqVector = new SequenceSelectionVector(0, 1, 5);
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I also see that this is sometimes different from test to test, so I would consider completely removing this `beforeEach` and simply initialize the same values in every test, I know this is somewhat repeating yourself but it help understand what each test it testing.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            seqVector.setIndex(0, 25);
+            seqVector.setIndex(2, -48);
+            seqVector.setIndex(3, 1000000000000001);
+
+            expect(seqVector.getIndex(0)).toBe(25);
+            expect(seqVector.getIndex(2)).toBe(-48);
+            expect(seqVector.getIndex(3)).toBe(1000000000000001);
+        });
+
+        it("Should materialize array on first set", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I'm not sure the title of this test is reflected in the test itself.
This is probably because materialize in an internal implementation details that is hard to reflect in the tests.
I would consider just making sure things are working as expected without "knowing" how they work internally.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            // After setting, array should be materialized
+            customVector.setIndex(2, 999);
+            expect(customVector.getIndex(2)).toBe(999);
+
+            // Other values should still be correct
+            expect(customVector.getIndex(0)).toBe(10);
+            expect(customVector.getIndex(1)).toBe(15);
+        });
+
+        it("Should return Index out of bounds", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: How is this different from the previous test?

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        });
+    });
+
+    describe("setLimit Test", () => {
+        it("Should set limit", () => {
+            seqVector.setLimit(250);
+            expect(seqVector.limit).toBe(250);
+            seqVector.setLimit(0);
+            expect(seqVector.limit).toBe(0);
+            seqVector.setLimit(-125);
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: How is this a valid limit?

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            expect(seqVector.limit).toBe(-125);
+        });
+
+        it("Should change limit independently of capacity", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I'm not sure I fully understand the difference between the two, it might be worth adding some comment in the code to explain this.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            const values = seqVector.selectionValues();
+            expect(values).toStrictEqual([0, 1, 2, 3, 4]);
+        });
+
+        it("Should return empty array for empty vector", () => {
+            const emptyVector = new SequenceSelectionVector(0, 1, 0);
+            expect(emptyVector.selectionValues()).toStrictEqual([]);
+        });
+
+        it("Should return correct values with custom base and delta", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: How is this different from the `Should return selectionVector` test?

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+            expect(negVector.selectionValues()).toStrictEqual([100, 90, 80, 70, 60]);
+        });
+
+        it("Should cache materialized array", () => {
+            const values1 = seqVector.selectionValues();
+            const values2 = seqVector.selectionValues();
+            expect(values1).toBe(values2); // Same reference
+        });
+
+        it("Should return materialized array after setIndex", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I think there's no need to mention the `materialized` part of it, the test simply makes sure that after setIndex the selectionValues still works as expected.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        });
+
+        it("Should initialize limit to size", () => {
+            const customVector = new SequenceSelectionVector(10, 5, 50);
+            expect(customVector.limit).toBe(50);
+            expect(customVector.capacity).toBe(50);
+        });
+    });
+
+    describe("Memory efficiency Test", () => {
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I think you'll have a hard time actually testing this.
Or to put it in other words, if the internal implementation would change these tests would still pass, so I don't think they are adding valuable information in terms of testing.
Consider removing them.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 6 above)
Comment (HarelM) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I now see the explanation, in order to "copy" the comment from the interface use `/** @inheritdoc */` (do check me) to say that this is the same as the interface it implements.

File: ts/src/vector/filter/selectionVectorUtils.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    return new FlatSelectionVector(selectionVector);
+}
+
+export function updateSelectionVector(selectionVector: SelectionVector, nullabilityBuffer: BitVector) {
+    const buffer = selectionVector.selectionValues();
+    let limit = 0;
+    for (let i = 0; i < selectionVector.limit; i++) {
+        if (!nullabilityBuffer || nullabilityBuffer.get(i)) {
+            buffer[limit++] = buffer[i];
```
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtils.ts]: Does this update the internal "memory" of the selectionVector? If so I think it's a bad idea...

File: ts/src/vector/filter/sequenceSelectionVector.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
+import { type SelectionVector } from "./selectionVector";
+
+export class SequenceSelectionVector implements SelectionVector {
+    private _baseValue: number;
```
Comment (HarelM) [inline ts/src/vector/filter/sequenceSelectionVector.ts]: You can use the same initialization as the `FlatVector` by using:
`constructor(private baseValue: number ...)`
I think it should be one or the other, but avoid using both, from a coding standard perspective.

File: ts/src/vector/filter/flatSelectionVector.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import { type SelectionVector } from "./selectionVector";
+
+export class FlatSelectionVector implements SelectionVector {
```
Comment (HarelM) [inline ts/src/vector/filter/flatSelectionVector.ts:7]: How is this different from a regular array? Because it implements `SelectionVector`? Do we really need this wrapper?

File: ts/src/vector/filter/flatSelectionVector.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        if (index >= this._limit || index < 0) {
+            throw new Error("Index out of bounds");
+        }
+
+        return this._selectionVector[index];
+    }
+
+    setIndex(index: number, value: number): void {
+        if (index >= this._limit || index < 0) {
+            throw new Error("Index out of bounds");
```
Comment (HarelM) [inline ts/src/vector/filter/flatSelectionVector.ts]: The other implementation uses `RangeError`, maybe it's worth using the same type of error?

File: ts/src/vector/filter/flatSelectionVector.spec.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
+import {beforeEach, describe, it, expect} from "vitest";
+import {FlatSelectionVector} from "./flatSelectionVector";
+
+let vector: number[];
+let FSVector: FlatSelectionVector;
```
Comment (HarelM) [inline ts/src/vector/filter/flatSelectionVector.spec.ts]: Use a different casing for this variable name (for example `fsVector`). This looks like a class name...

File: ts/src/vector/filter/selectionVectorUtil.spec.ts
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtil.spec.ts:1]: There might be a bit too many tests in this file.
The code in the utils should be tested without testing the code of `FlatSelectionVector` or `SequenceSelectionVector` as much as possible.
This test files feels like it repeats some of the other tests, which I'm not sure it necessary.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 1 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: see next comment

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 2 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: Each test now initializes its own vector, making it clear what values each test validates.

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 3 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: Changed test names in general to reflect the expected outcome better

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 5 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I added a check in the setLimit method

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 4 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I changed the index test setup to check get and set Limit together

File: ts/src/vector/filter/sequenceSelectionVector.spec.ts (hunk 6 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.spec.ts]: I have added the /** @inheritdoc */

File: ts/src/vector/filter/flatSelectionVector.spec.ts (hunk 14 above)
Comment (Salkin975) [reply ts/src/vector/filter/flatSelectionVector.spec.ts]: I changed the name

File: ts/src/vector/filter/flatSelectionVector.ts (hunk 13 above)
Comment (Salkin975) [reply ts/src/vector/filter/flatSelectionVector.ts]: Changed the error to also be of type RangeError

File: ts/src/vector/filter/selectionVectorUtils.ts (hunk 10 above)
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtils.ts]: It did, but I changed it to not use it anymore

File: ts/src/vector/filter/sequenceSelectionVector.ts (hunk 11 above)
Comment (Salkin975) [reply ts/src/vector/filter/sequenceSelectionVector.ts]: I created the constructor

File: ts/src/vector/filter/selectionVectorUtil.spec.ts
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtil.spec.ts:1]: I have removed unnecessary tests

File: ts/src/vector/filter/selectionVectorUtils.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    for (let i = 0; i < selectionVector.limit; i++) {
+        if (!nullabilityBuffer || nullabilityBuffer.get(i)) {
+            filteredIndices.push(selectionVector.getIndex(i));
+        }
+    }
+    return new FlatSelectionVector(filteredIndices);
+}
+
+/** Returns new selection vector with only indices where nullability is true. */
+export function updateNullableSelectionVector(selectionVector: SelectionVector, nullabilityBuffer: BitVector): SelectionVector {
```
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtils.ts]: Isn't this and the method above identical?

File: ts/src/vector/filter/selectionVectorUtils.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    const filteredIndices = [];
+    for (let i = 0; i < selectionVector.limit; i++) {
+        if (!nullabilityBuffer || nullabilityBuffer.get(i)) {
```
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtils.ts]: Are you sure about this condition? It looks reversed...

File: ts/src/vector/filter/flatSelectionVector.ts (hunk 12 above)
Comment (Salkin975) [reply ts/src/vector/filter/flatSelectionVector.ts:7]: It was originally the only SelectionVector implementation. I added the sequenceSelectionVector as part of a todo (https://github.com/Turtelll/maplibre-tile-spec/blob/a8edfa36480d32fd5fd17770bb3974b5fda8c7fc/ts/src/vector/constant/intConstVector.ts#L17), so I don't think it should be removed
https://github.com/Turtelll/maplibre-tile-spec/tree/a8edfa36480d32fd5fd17770bb3974b5fda8c7fc/ts/src/vector/filter

File: ts/src/vector/filter/selectionVectorUtils.ts (hunk 15 above)
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtils.ts]: yes they are, I will remove one

File: ts/src/vector/filter/selectionVectorUtils.ts (hunk 16 above)
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtils.ts]: it used the wrong index, I updated it to use the actual Index of the selectionVector

File: ts/src/vector/filter/flatSelectionVector.ts (hunk 12 above)
Comment (HarelM) [reply ts/src/vector/filter/flatSelectionVector.ts:7]: Fair enough.

File: ts/src/vector/filter/selectionVectorUtils.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    return new FlatSelectionVector(selectionVector);
+}
+
+/** Returns new selection vector with only indices where nullability is true. */
+export function updateNullableSelectionVector(selectionVector: SelectionVector, nullabilityBuffer: BitVector): SelectionVector {
+    const filteredIndices = [];
+    for (let i = 0; i < selectionVector.limit; i++) {
+        const vectorIndex = selectionVector.getIndex(i);
+        if (!nullabilityBuffer || nullabilityBuffer.get(vectorIndex)) {
```
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtils.ts:36]: Doesn't `!nullabilityBuffer` means that if `nullabilityBuffer` is defined (not null) it will add all the values of the original selection vector? I'm still confused.
I did look at the tests, and they seems to test this, but I'm still finding this code confusing, what am I missing?

File: ts/src/vector/filter/selectionVectorUtils.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+import { SequenceSelectionVector } from "./sequenceSelectionVector";
+
+export function createSelectionVector(size: number) {
+    return new SequenceSelectionVector(0, 1, size);
+}
+
+export function createNullableSelectionVector(size: number, nullabilityBuffer: BitVector) {
+    const selectionVector = [];
+    for (let i = 0; i < size; i++) {
+        if (nullabilityBuffer?.get(i)) {
```
Comment (HarelM) [inline ts/src/vector/filter/selectionVectorUtils.ts]: If `nullabilityBuffer` is not passed it will return an empty array, wouldn't it?
It is the expected results? I haven't seen a test to cover this input.

File: ts/src/vector/filter/flatSelectionVector.ts [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    setIndex(index: number, value: number): void {
+        if (index >= this._limit || index < 0) {
+            throw new RangeError("Index out of bounds");
+        }
+
+        this._selectionVector[index] = value;
+    }
+
+    /** @inheritdoc */
+    setLimit(limit: number): void {
```
Comment (HarelM) [inline ts/src/vector/filter/flatSelectionVector.ts:41]: I think this method is missing coverage...

File: ts/src/vector/filter/selectionVectorUtils.ts (hunk 17 above)
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtils.ts:36]: I didn't implement the method myself, but I understand it as:
No nullability buffer = all values are considered NOT null
With nullability buffer = only values where the bit is 1 are valid

Comment (HarelM) [reply ts/src/vector/filter/selectionVectorUtils.ts:36]: Yeah, you are right, `!nullabilityBuffer` is equal to `nullabilityBuffer == null`, so all values will be added.

**Comment by HarelM** [comment]
Thanks for all the work on this!
> ✅ All modified and coverable lines are covered by tests.

Is a great indication of good coverage!
There's only one last comment remaining before merging this, unless you would like to wait with it for other parts to be added?

File: ts/src/vector/filter/selectionVectorUtils.ts (hunk 18 above)
Comment (Salkin975) [reply ts/src/vector/filter/selectionVectorUtils.ts]: I have added comments to both methods and changed them to use the same logic. I also added tests for these edge cases

File: ts/src/vector/filter/flatSelectionVector.ts (hunk 19 above)
Comment (Salkin975) [reply ts/src/vector/filter/flatSelectionVector.ts:41]: is done

**Comment by Salkin975** [comment]
No, these changes are required for the other ones
